### ISSUE TITLE
hang from duff prototype

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13547]
Migrated from [rt.perl.org#121050](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=121050) (status was 'resolved')

Searchable as RT121050$

**Comment by p5pRT** [comment]
### From zefram@fysh.org

#### Created by zefram@fysh.org



$ cat t0
sub foo \( 
\)
\{ print 123; \}
foo\(\);
$ perl5\.19\.7 t0

The above causes perl to hang\.  It's behaved like that since 5\.19\.2\.
Earlier perls show different interpretations of the code&#8203;:

$ perl5\.19\.1 \-MO=Deparse t0
sub foo \(\) \{
&nbsp;    print 123;
\}
foo;
t0 syntax OK
$ perl5\.17\.4 \-MO=Deparse t0
sub foo \(\) \{
&nbsp;    print 123;
\}
foo;
t0 syntax OK
$ perl5\.17\.3 \-MO=Deparse t0
sub print \(\) \{
&nbsp;    print 123;
\}
foo\(\);
t0 syntax OK

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.19.7:

Configured by zefram at Mon Dec 23 19:18:49 GMT 2013.

Summary of my perl5 (revision 5 version 19 subversion 7) configuration:
   
  Platform:
    osname=linux, osvers=3.2.0-4-amd64, archname=x86_64-linux-thread-multi
    uname='linux barba.rous.org 3.2.0-4-amd64 #1 smp debian 3.2.46-1 x86_64 gnulinux '
    config_args='-des -Dprefix=/home/zefram/usr/perl/perl_install/perl-5.19.7-i64-f52 -Duselargefiles -Dusethreads -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dusedevel -Uversiononly -Ui_db'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.7.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib /usr/lib
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.15'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E -Wl,-rpath,/home/zefram/usr/perl/perl_install/perl-5.19.7-i64-f52/lib/5.19.7/x86_64-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'



@INC for perl 5.19.7:
    /home/zefram/usr/perl/perl_install/perl-5.19.7-i64-f52/lib/site_perl/5.19.7/x86_64-linux-thread-multi
    /home/zefram/usr/perl/perl_install/perl-5.19.7-i64-f52/lib/site_perl/5.19.7
    /home/zefram/usr/perl/perl_install/perl-5.19.7-i64-f52/lib/5.19.7/x86_64-linux-thread-multi
    /home/zefram/usr/perl/perl_install/perl-5.19.7-i64-f52/lib/5.19.7
    .


Environment for perl 5.19.7:
    HOME=/home/zefram
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/zefram/usr/perl/perl_install/perl-5.19.7-i64-f52/bin:/home/zefram/usr/perl/util:/home/zefram/pub/x86_64-unknown-linux-gnu/bin:/home/zefram/pub/common/bin:/usr/bin:/bin:/usr/local/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From perl5-porters@perl.org

Zefram wrote&#8203;:
> The above causes perl to hang\.  It's behaved like that since 5\.19\.2\.
> Earlier perls show different interpretations of the code&#8203;:

The 5\.18 behaviour \(same as 5\.17\.4\) makes the most sense\, so this is a
regression\.  I think it ought to be a 5\.20 blocker\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @wolfsage

I can't reproduce this with blead\. Are there special characters in the
example I'm not seeing or is something else going on?

\-\- Matthew Horsfall \(alh\)

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Matthew Horsfall \(alh\) wrote&#8203;:
>I can't reproduce this with blead\. Are there special characters in the
>example I'm not seeing or is something else going on?

No special characters\.  As there's obviously some memory corruption
involved\, it's likely to be platform dependent\, which is the most likely
reason for failure to reproduce\.  On one platform it may well also be
affected by config options\.

\-zefram

**Comment by p5pRT** [comment]
### From @wolfsage

On Fri\, Jan 24\, 2014 at 9&#8203;:00 AM\, Zefram \<zefram@&#8203;fysh\.org> wrote&#8203;:
> Matthew Horsfall \(alh\) wrote&#8203;:
>>I can't reproduce this with blead\. Are there special characters in the
>>example I'm not seeing or is something else going on?
>
> No special characters\.  As there's obviously some memory corruption
> involved\, it's likely to be platform dependent\, which is the most likely
> reason for failure to reproduce\.  On one platform it may well also be
> affected by config options\.

Hrm\, alright thank you\.

I did attempt building with your config options to see if that would
do it but no luck\.

\-\- Matthew Horsfall \(alh\)

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed in c4cf781e24a\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
